core/arch/aarch64/codec.txt

```diff
@@ -1128,3 +1132,58 @@ x101101011000000000101xxxxxxxxxx  cls     wx0 : wx5
 00000100xx1xxxxx000101xxxxxxxxxx     uqadd     z0 : z5 z16 bhsd_sz
 00000100xx1xxxxx000110xxxxxxxxxx     sqsub     z0 : z5 z16 bhsd_sz
 00000100xx1xxxxx000111xxxxxxxxxx     uqsub     z0 : z5 z16 bhsd_sz
+
+# Advanced SIMD three different
```

Comment:
I think we could do with a comment explaining what the width operand means in this encoding group. 

E.g. for the *L variants, the width of the inputs can be byte, half and single, and the output width will be half, single and double respectively. In that case, we use the size of the input operands for the width operand, which in turn implies the size of the output operand. The other variants in the group are handled in a similar fashion. In the actual encoding, the size of the source operands is used as well.

@egrimley @derekbruening it would be great if you could have a quick look and see if that approach makes sense to you.

Reply:
> E.g. for the *L variants, the width of the inputs can be byte, half and single, and the output width will be half, single and double respectively. In that case, we use the size of the input operands for the width operand, which in turn implies the size of the output operand. The other variants in the group are handled in a similar fashion. In the actual encoding, the size of the source operands is used as well.

I recall we had a conversation on whether to put this into the opcode (which would better match x86) but the decision was to go with a special immediate operand to augment the opcode, so this is just matching the earlier decision, right?

Reply:
AFAIK we do not have any use case or client in mind while doing this work. We mostly are just trying to get rid of OP_xx at some point.

As you said, different representations might be slightly more convenient for different use cases, but I think in any case the client needs special logic to interpret the size immediate(s). The advantage of the current representation is that it is closer to the actual binary encoding, as the a single size is encoded in the 'size' bits.

If we encounter encodings the require 2 different sizes in the binary encoding, having multiple immediates seems quite a natural extension.

Reply:
Sure, works for me.